Added a transpile-only ESM loader (#1101)

concision opened the pull request:
This resolves the feature request #1101: A new ESM loader equivalent of 'ts-node/register/transpile-only'

cspotcode commented:
Looks good but it'll need tests before being merged, and our package.json will need to be updated.

You can probably copy-paste an existing ESM test, but add a deliberate type error to make sure it gets ignored.

We also have tests that each entrypoint can be resolved, which will validate the changes to package.json.

concision commented:
I have added `esm/transpile-only.mjs` to the `package.json` exports and corresponding resolution tests.

I also added two tests to verify the loader should behave as expected:
- verify that the transpile-only ESM loader skips type checks in the presence of type errors
- verify that the transpile-only ESM loader throws a type error in the presence of type errors

Are there any other needed changes I need to make?

cspotcode commented on tests/esm-transpile-only/index.ts:
```diff
@@ -0,0 +1,12 @@
+import { foo } from './foo.js'
+import { bar } from './bar.js'
+import { baz } from './baz.js'
+import { biff } from './biff.js'
```
Do you think we can remove the foo, bar, baz, and biff files?  Since `esm-transpile-only` is focused on verifying the transpile-only loader, so the focus is on the semantic errors.

concision replied:
Sounds good; removed them.

cspotcode commented on src/index.spec.ts:
```diff
@@ -802,6 +804,31 @@ describe('ts-node', function () {
           return done()
         })
       })
+
```
Nitpick: this suite name might imply that the ESM loader only supports transpile-only via this entrypoint, though it can also load this flag from environment variables or tsconfig.json.

Maybe name the 2x test cases to match some of our other specs' names, and remove the `describe()` wrapper?
```
it('should support transpile only mode via dedicated loader entrypoint', ...
it('should throw errors', ...
```

concision replied:
Is this acceptable?
https://github.com/TypeStrong/ts-node/blob/525b7308fc45355fa46f6306c36895e6de19c2fd/src/index.spec.ts#L808-L830

cspotcode replied:
Yeah, looks good

cspotcode commented:
@concision looks great, thanks.  It's merged.  I need a review of #970 before making our next release, but hopefully we can get that done this weekend.  It'll be 9.0.0
